### PR TITLE
Merge distance based speckle filter with RadiusOutlier removal

### PR DESCRIPTION
**Pull request by nlimpert**
As discussed in https://github.com/ros-perception/laser_filters/pull/97 merge the distance based speckle filter with the RadiusOutlier filter.

Note that due to the merging of the parameter sets I removed the `out_of_range` check in c9ed9e1c9b5869e5be72f1a8b6a2553f92d39a8e.
Please let me know if that was a mistake.

### CODE REVIEW

**Comment by nlimpert** [comment]
@reinzor fyi

**Comment by reinzor** [comment]
Nice work. Regarding the out of range check. I think we would like to configure this filter for a specific max range since often these so called speckles occur when the laser scanner is close to obstacles. You want to filter out these cases but you don't want to remove 1 point measurements far away since they might be valid.

I will quickly go through your PR and test this hopefully next week (depends a little bit on the corona situation).

File: src/speckle_filter.cpp [hunk 1]
```diff
@@ -117,5 +103,27 @@ bool LaserScanSpeckleFilter::update(const sensor_msgs::LaserScan& input_scan, se
 void LaserScanSpeckleFilter::reconfigureCB(laser_filters::SpeckleFilterConfig& config, uint32_t level)
 {
   config_ = config;
+
+  switch (config_.filter_type) {
+    case laser_filters::SpeckleFilter_RadiusOutlier:
+      if (!validator_)
```
Comment (reinzor) [inline src/speckle_filter.cpp:109]: shouldn't this be: `if (validator_)`?

File: src/speckle_filter.cpp [hunk 2]
```diff
@@ -117,5 +103,27 @@ bool LaserScanSpeckleFilter::update(const sensor_msgs::LaserScan& input_scan, se
...
+      {
+        delete validator_;
+      }
+      validator_ = new laser_filters::RadiusOutlierWindowValidator();
+      break;
+
+    case laser_filters::SpeckleFilter_Distance:
+      if (!validator_)
```
Comment (reinzor) [inline src/speckle_filter.cpp:117]: shouldn't this be: if (validator_)?

Comment (nlimpert) [reply src/speckle_filter.cpp]: Of course you're correct! Thanks for the heads up! Done in 61828e6ef821943f2a0cb02366dcb9800dd99e0a

File: src/speckle_filter.cpp (hunk 1 above)
Comment (nlimpert) [reply src/speckle_filter.cpp]: Of course you're correct! Thanks for the heads up! Done in a1b1ae3ee39bfa93386ba726ec8541e95e7911e9

**Comment by nlimpert** [comment]
> Nice work. Regarding the out of range check. I think we would like to configure this filter for a specific max range since often these so called speckles occur when the laser scanner is close to obstacles. You want to filter out these cases but you don't want to remove 1 point measurements far away since they might be valid.
> 
> I will quickly go through your PR and test this hopefully next week (depends a little bit on the corona situation).

Thanks for taking a look! I re-introduced the out of range check with an additional parameter to be set.

**Comment by reinzor** [comment]
I think this is more clear indeed. I will try to test this tomorrow on our robots.

**Comment by nlimpert** [comment]
I've noticed that I forgot to adapt the parameters of the test. Fixed in cc2f961.
Sorry for the spam :-)

**Comment by reinzor** [comment]
Shouldn't we create this PR in the `ros` repo btw, this is a fork?

**Comment by nlimpert** [comment]
> Shouldn't we create this PR in the `ros` repo btw, this is a fork?

The base branch of my PR in your fork is `eurogroep:feat/speckle-filter-for-noise-removal` which, in turn, is the target branch of https://github.com/ros-perception/laser_filters/pull/97.
So once we discussed my changes here and you'll merge my PR here it will be merged in `eurogroep:feat/speckle-filter-for-noise-removal` and the upstream PR should be updated automatically.
We'd probably only have to notify the upstream maintainer on the status then.

**Comment by reinzor** [comment]
Tested on our robots. Behavior seems the same as "old" speckle filter.

**Pull request merged**